flux/src/main/java/software/amazon/aws/clients/swf/flux/poller/DecisionTaskPoller.java

```diff
@@ -576,9 +589,9 @@ static RespondDecisionTaskCompletedRequest decide(Workflow workflow, WorkflowSte
             // Note that we can't get here if there aren't any successfully-scheduled partitions, since we would have
             // generated partition metadata and returned from this method earlier.
 
-            List<PartitionState> states = state.getStepPartitions().get(nextActivityName)
-                    .values().stream().filter(ps -> !ps.isEmpty()).findFirst().orElse(Collections.emptyList());
-            PartitionState lastState = states.get(states.size() - 1);
+            PartitionState lastState = state.getLatestPartitionStates(nextActivityName)
+                    .values().stream().filter(Objects::nonNull).findFirst().orElse(null);
+            // TODO -- Is it possible for the .orElse() in the previous line to happen? If so, this will throw.
```

Comment:
This kinda feels like something we need to know before we make this change, no?

Reply:
And this same sort of question exists for a similar block of code at the top of this PR

Reply:
Yeah, I copied the TODO because I didn't want to think through whether it was still relevant, but I'll stop being lazy and write it out here.

In this code, we're generating input for the *next* workflow step to execute.

`state.getLatestPartitionStates(stepName)` returns a `Map<String, PartitionState>` containing one entry for each successfully-scheduled partition for the step named `stepName`, or `null` if there has never been a successfully-scheduled attempt but partition metadata was generated.

The map is always non-null because `getLatestPartitionStates()` returns `Collections.emptyMap()` in that case.

If the map is empty, then there were never any successfully-scheduled attempts for any partitions *and* no partition metadata marker was generated. In this case, the code on lines 512-518 will generate a partition metadata marker and we won't get to this line here.

If the map is not empty but contains only `null` values, then no attempts for *any* partition were successfully scheduled,. However, the only way for us to have the partition IDs without having any successfully-scheduled attempts is for us to have previously generated a metadata marker, in which case we don't enter this if block at all, since line 585 will handle it.

If the map contains at least one non-null value, then the `.orElse()` block referenced by the TODO will never trigger.

---

For the instance of the same TODO comment on line 354, the reasoning is as follows:

There is a null check against the result of `state.getLatestPartitionStates(stepName)` immediately prior. The step being looked up is the *most recent step*, and the only way we won't have state for the most recent step is if the workflow has just begun, in which case `currentStep` is null and the previous if block (line 336) handles it.

---

I'll go ahead and remove these TODOs and replace them with notes explaining the above.